### DISCUSSION TITLE
config.cookie_enabled = true doesn't work?

### DISCUSSION THREAD

**Discussion by meliborn** [graphql-devise/graphql_devise#224] [Q&A]
Based on https://devise-token-auth.gitbook.io/devise-token-auth/config/initialization I was trying to send auth toke via Cookie, but it doesn't work. Am I missed something?

**Comment by 00dav00** [comment]
Hi @meliborn , do you have more details about the issue you are experiencing? Perhaps the backtrace of the error?

**Reply by meliborn** [reply]
@00dav00 there is no error, I just added `config.cookie_enabled = true` to devise token initializer, restarted the app and don't see any cookies in response after successful login. It keeps send them via headers

**Reply by mcelicalderon** [reply]
That is weird, @meliborn. Looking at the method we use to check if a user is signed in https://github.com/lynndylanhurley/devise_token_auth/blob/798255ee7e3fee5cfa2fdc519fd90e281bd9b6f5/app/controllers/devise_token_auth/concerns/set_user_by_token.rb#L40 it should work if the correct cookie is set in the request. I'm pretty sure (I'd have to confirm) that our login mutation does set the cookie when you login (by calling devise's `controller.sign_in`) as I even remember we had a bug where the cookie would overflow.

Could you please confirm that when making the login request you are getting the cookie in the response? And that when you make a request after that you are settig the right cookie in the request?

**Reply by meliborn** [reply]
@00dav00 that's the problem - I do not getting the cookie in the response. I'm using Insomnia to test Graphql requests. Maybe there is a problem with cors\cookie policy or something like that? But I remember I did testing with cookies in previous similar project with the same conditions, but without `graphql_devise` gem. 
By design, should enabling cookie prevents token being sent in headers?

**Reply by meliborn** [reply]
@00dav00 from what I understood, `controller.sign_in` doesn't set the cookie. It could only set `session_id`, but only if `store: true` being passed, which is doesn't.

**Reply by 00dav00** [reply]
Hi @meliborn , I just had a look at what is happening here. This seems to be an issue with gql_devise not being ready to handle auth tokens via cookies.

This is [the method](https://github.com/lynndylanhurley/devise_token_auth/blob/master/app/controllers/devise_token_auth/concerns/set_user_by_token.rb#L96) DTA uses to set the auth headers/cookies. As you can see `cookie_enabled` is considered when updating the creds.
This is [the method](https://github.com/graphql-devise/graphql_devise/blob/f648b6a281d6f8e40fca41f135a2e82f42537563/lib/graphql_devise/concerns/controller_methods.rb#L69) gql_devise uses to set the auth headers, in here DTA's config about cookies is not considered. 

We will have to add this into the gem in order to support cookies properly. I don't think I can provide you with a short cut for the time being but this [part of DTA's code](https://github.com/lynndylanhurley/devise_token_auth/blob/master/app/controllers/devise_token_auth/concerns/set_user_by_token.rb#L103-L116) can guide on how you would need to implement cookie support manually until we find some time to add it into the gem.

**Reply by mcelicalderon** [reply]
@meliborn @00dav00 I'm pretty sure the gem should be able to handle cookies correctly. We fixed this in the past https://github.com/graphql-devise/graphql_devise/issues/112 and I think it might be just a configuration problem, perhaps the base controller you are using. I need to do some testing, but if you could provide more details about how you are using this gem, that would help. Are you mounting in your own schema or using a separate route? Is your graphql controller api only?

**Reply by meliborn** [reply]
@It was mounted in the existing schema. Controller isn't an api only.

**Reply by 00dav00** [reply]
@mcelicalderon  The cookie referenced in #112 is the one [set by Warden](https://github.com/heartcombo/devise/blob/main/lib/devise/controllers/sign_in_out.rb#L68). The one DTA uses is [this one](https://github.com/lynndylanhurley/devise_token_auth/blob/798255ee7e3fee5cfa2fdc519fd90e281bd9b6f5/app/controllers/devise_token_auth/concerns/set_user_by_token.rb#L152) which is set in an [after action callback](https://github.com/lynndylanhurley/devise_token_auth/blob/798255ee7e3fee5cfa2fdc519fd90e281bd9b6f5/app/controllers/devise_token_auth/concerns/set_user_by_token.rb#L9) by DTA's `SetUserByToken` concern.

gql_devise concern `SetUserByToken` does not implement the callbacks that set this cookie, the only method wrapped from DTA is [set_user_by_token](https://github.com/graphql-devise/graphql_devise/blob/master/lib/graphql_devise/concerns/additional_controller_methods.rb#L35) which works to identify the user during the initial part the request.

I think this needs to be implemented. Could you confirm please?

**Reply by 00dav00** [reply]
Hi @meliborn , the change on [this PR](https://github.com/graphql-devise/graphql_devise/pull/233) implements the changes needed to set the auth cookies when you have this option configured in DTA.

It is important to consider that setting auth cookies will only work if you are using the schema in  a GQL route within your main rails project. For now the route mounted using `mount_graphql_devise_for` is inhering from `ActionController::API` and can't be configured to inherit from a controller that includes the rails cookie concern.

This change is in master now and we will be releasing a new version of the next days.

**Reply by meliborn** [reply]
@00dav00 glad to hear!
